File: frontend/apps/ios/Pantopus/Features/Homes/VerifyLandlord/VerifyLandlordWizardViewModel.swift
//
//  VerifyLandlordWizardViewModel.swift
//  Pantopus
//
//  Drives the A12.5 / A12.6 wizard state machine:
//
//    .start → .details → submit → openPostcardVerification(homeId)
//
//  On submit it requests the verification postcard via
//  `POST /api/homes/:id/request-postcard` (route
//  `backend/routes/homeOwnership.js:2452`) then dispatches the outbound
//  `openPostcardVerification` event. The landlord/PM details collected
//  in the form have no backend representation (request-postcard takes no
//  body), so they stay client-side.
//

import Foundation
import Observation

/// View model backing `VerifyLandlordWizardView`. Holds the per-field
/// form state, the current step, and the submit state machine
/// (`.idle → .submitting → .submitted / .error(_)`).
@Observable
@MainActor
final class VerifyLandlordWizardViewModel: WizardModel {
    // MARK: - Published state

    private(set) var currentStep: VerifyLandlordStep = .start
    private(set) var startContent: VerifyLandlordStartContent
    var form: VerifyLandlordForm
    /// Validation errors materialised lazily — `nil` means "user hasn't
    /// tried to submit yet, don't render error chips". Becomes
    /// `.empty` or populated after the first submit attempt.
    private(set) var errors: VerifyLandlordValidationErrors?
    private(set) var submitState: VerifyLandlordSubmitState = .idle
    var pendingEvent: VerifyLandlordOutboundEvent?

    // MARK: - Init

    private let homeId: String
    private let submitDelayNanos: UInt64
    private let api: APIClient

    /// Test/offline seam for the postcard request. When non-nil,
    /// `submit()` calls this instead of
    /// `POST /api/homes/:id/request-postcard`.
    public typealias PostcardRequester = @MainActor () async -> Result<Void, any Error>
    private let postcardRequester: PostcardRequester?

    init(
        homeId: String,
        startContent: VerifyLandlordStartContent? = nil,
        form: VerifyLandlordForm? = nil,
        api: APIClient = .shared,
        submitDelayNanos: UInt64 = 800_000_000,
        postcardRequester: PostcardRequester? = nil
    ) {
        self.homeId = homeId
        self.startContent = startContent
            ?? VerifyLandlordSampleData.startContent(for: homeId)
        self.form = form ?? VerifyLandlordSampleData.formSeed(for: homeId)
        self.api = api
        self.submitDelayNanos = submitDelayNanos
        self.postcardRequester = postcardRequester
    }

    // MARK: - WizardModel

    var chrome: WizardChrome {
        let dirty = !form.ownerName.isEmpty
            || !form.contactName.isEmpty
            || !form.email.isEmpty
            || form.lease != nil
            || form.pmEnabled
        switch currentStep {
        case .start:
            return WizardChrome(
                title: "Verify landlord",
                progressLabel: .stepOf(current: 1, total: 3),
                progressFraction: 1.0 / 3.0,
                leading: .close,
                primaryCTALabel: "Start verification",
                primaryCTAEnabled: true,
                secondaryCTA: nil,
                isSubmitting: false,
                dirty: dirty,
                showsProgressBar: true
            )
        case .details:
            let live = form.validate()
            let blocked = (errors != nil && !live.isEmpty) || isSubmitting
            return WizardChrome(
                title: "Verify landlord",
                progressLabel: .stepOf(current: 2, total: 3),
                progressFraction: 2.0 / 3.0,
                leading: .back,
                primaryCTALabel: "Submit",
                primaryCTAEnabled: !blocked,
                secondaryCTA: nil,
                isSubmitting: isSubmitting,
                dirty: dirty,
                showsProgressBar: true
            )
        }
    }

    var isSubmitting: Bool {
        if case .submitting = submitState { return true }
        return false
    }

    func leadingTapped() {
        switch currentStep {
        case .start:
            pendingEvent = .dismiss
        case .details:
            currentStep = .start
            errors = nil
        }
    }

    func discardConfirmed() {
        pendingEvent = .dismiss
    }

    func primaryTapped() {
        switch currentStep {
        case .start:
            currentStep = .details
        case .details:
            Task { await submit() }
        }
    }

    func secondaryTapped() {}

    // MARK: - Form mutations

    func setOwnerName(_ value: String) {
        form.ownerName = value
        refreshErrorsIfShown()
    }

    func setContactName(_ value: String) {
        form.contactName = value
        refreshErrorsIfShown()
    }

    func setEmail(_ value: String) {
        form.email = value
        refreshErrorsIfShown()
    }

    func setPhone(_ value: String) {
        form.phone = value
    }

    func setLease(_ lease: VerifyLandlordLeaseFile?) {
        form.lease = lease
        refreshErrorsIfShown()
    }

    func setPMEnabled(_ enabled: Bool) {
        form.pmEnabled = enabled
        if !enabled {
            form.pmName = ""
            form.pmEmail = ""
            form.pmPhone = ""
        }
        refreshErrorsIfShown()
    }

    func setPMName(_ value: String) {
        form.pmName = value
        refreshErrorsIfShown()
    }

    func setPMEmail(_ value: String) {
        form.pmEmail = value
        refreshErrorsIfShown()
    }

    func setPMPhone(_ value: String) {
        form.pmPhone = value
    }

    // MARK: - Submit

    func submit() async {
        if isSubmitting { return }
        let live = form.validate()
        errors = live
        if !live.isEmpty {
            submitState = .error(message: "Fix \(live.count) thing\(live.count == 1 ? "" : "s") to submit")
            return
        }
        submitState = .submitting
        if !NetworkMonitor.shared.isOnline {
            submitState = .error(message: "You're offline. Try again when you're back online.")
            return
        }
        // The landlord / property-manager details collected above have no
        // backend representation today — `request-postcard` takes no body
        // — so finishing the wizard simply mails the verification postcard
        // and routes to the code-entry screen. The gathered details stay
        // client-side (documented follow-up).
        switch await requestPostcard() {
        case .success:
            submitState = .submitted
            pendingEvent = .openPostcardVerification(homeId: homeId)
        case let .failure(error):
            // A pending/duplicate code (400) or address cap (429) means a
            // postcard is already on its way — proceed to enter it. Other
            // failures surface inline so the user can retry.
            if case let .clientError(status, _) = (error as? APIError), status == 400 || status == 429 {
                submitState = .submitted
                pendingEvent = .openPostcardVerification(homeId: homeId)
            } else {
                submitState = .error(
                    message: (error as? APIError)?.errorDescription
                        ?? "Couldn't request the verification postcard. Try again."
                )
            }
        }
    }

    /// Mails the verification postcard. Uses the injected
    /// `postcardRequester` seam when present (previews/tests); otherwise
    /// calls `POST /api/homes/:id/request-postcard`.
    private func requestPostcard() async -> Result<Void, any Error> {
        if let postcardRequester {
            try? await Task.sleep(nanoseconds: submitDelayNanos)
            return await postcardRequester()
        }
        do {
            _ = try await api.request(
                HomesEndpoints.requestPostcard(homeId: homeId),
                as: RequestPostcardResponse.self
            )
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    func acknowledgePendingEvent() {
        pendingEvent = nil
    }

    // MARK: - Variant switching

    // Used by previews / sample data toggles + the dashboard fast-track decision tree.

    func setVariant(_ variant: VerifyLandlordVariant) {
        guard variant != startContent.variant else { return }
        switch variant {
        case .canonical: startContent = VerifyLandlordSampleData.canonical
        case .fastTrack: startContent = VerifyLandlordSampleData.fastTrack
        }
    }

    // MARK: - Helpers

    /// Re-run validation only after the user has already attempted
    /// submit once — so we don't spray error chips while they're still
    /// typing the email.
    private func refreshErrorsIfShown() {
        guard errors != nil else { return }
        errors = form.validate()
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/verify_landlord/VerifyLandlordWizardViewModel.kt
@file:Suppress("MagicNumber", "PackageNaming", "TooManyFunctions")

package app.pantopus.android.ui.screens.homes.verify_landlord

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomeVerificationRepository
import app.pantopus.android.data.network.NetworkMonitor
import app.pantopus.android.ui.screens.shared.wizard.WizardChrome
import app.pantopus.android.ui.screens.shared.wizard.WizardLeadingControl
import app.pantopus.android.ui.screens.shared.wizard.WizardModel
import app.pantopus.android.ui.screens.shared.wizard.WizardProgressLabel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Nav-arg key for the home being verified. */
const val VERIFY_LANDLORD_HOME_ID_KEY: String = "homeId"

/**
 * Aggregate UI state for the verify-landlord wizard. Combined into a
 * single record so the screen derives [WizardChrome] from one state
 * read.
 */
data class VerifyLandlordUiState(
    val currentStep: VerifyLandlordStep = VerifyLandlordStep.Start,
    val startContent: VerifyLandlordStartContent = VerifyLandlordSampleData.canonical,
    val form: VerifyLandlordForm = VerifyLandlordForm(),
    /**
     * Validation errors materialised lazily — `null` means "user
     * hasn't tried to submit yet, don't render error chips". Becomes
     * an empty [VerifyLandlordValidationErrors] or populated after the
     * first submit attempt.
     */
    val errors: VerifyLandlordValidationErrors? = null,
    val submitState: VerifyLandlordSubmitState = VerifyLandlordSubmitState.Idle,
) {
    val isSubmitting: Boolean get() = submitState is VerifyLandlordSubmitState.Submitting

    val isDirty: Boolean
        get() =
            form.ownerName.isNotEmpty() ||
                form.contactName.isNotEmpty() ||
                form.email.isNotEmpty() ||
                form.lease != null ||
                form.pmEnabled
}

/**
 * Drives the A12.5 / A12.6 wizard state machine:
 *
 *     Start -> Details -> submit -> OpenPostcardVerification(homeId)
 *
 * On submit it requests the verification postcard via
 * `POST /api/homes/:id/request-postcard` (route
 * `backend/routes/homeOwnership.js:2452`) then dispatches the outbound
 * `OpenPostcardVerification` event. The landlord/PM details collected in
 * the form have no backend representation (request-postcard takes no
 * body), so they stay client-side.
 */
@HiltViewModel
open class VerifyLandlordWizardViewModel
    @Inject
    constructor(
        private val networkMonitor: NetworkMonitor,
        savedStateHandle: SavedStateHandle,
        private val verificationRepository: HomeVerificationRepository,
    ) : ViewModel(),
        WizardModel {
        private val homeId: String =
            requireNotNull(savedStateHandle[VERIFY_LANDLORD_HOME_ID_KEY]) {
                "VerifyLandlordWizardViewModel requires a '$VERIFY_LANDLORD_HOME_ID_KEY' nav arg."
            }

        /** Configurable so JVM unit tests can drop the delay to zero. */
        protected open val submitDelayMillis: Long = SUBMIT_DELAY_DEFAULT_MILLIS

        private val _state =
            MutableStateFlow(
                VerifyLandlordUiState(
                    startContent = VerifyLandlordSampleData.startContent(homeId),
                    form = VerifyLandlordSampleData.formSeed(homeId),
                ),
            )
        val state: StateFlow<VerifyLandlordUiState> = _state.asStateFlow()

        /** One-shot navigation events the screen reacts to. */
        val pendingEvent = MutableStateFlow<VerifyLandlordOutboundEvent?>(null)

        // MARK: - WizardModel

        override val chrome: WizardChrome
            get() = computeChrome(_state.value)

        override fun onLeading() {
            when (_state.value.currentStep) {
                VerifyLandlordStep.Start -> pendingEvent.value = VerifyLandlordOutboundEvent.Dismiss
                VerifyLandlordStep.Details -> {
                    _state.update { it.copy(currentStep = VerifyLandlordStep.Start, errors = null) }
                }
            }
        }

        override fun onDiscard() {
            pendingEvent.value = VerifyLandlordOutboundEvent.Dismiss
        }

        override fun onPrimary() {
            when (_state.value.currentStep) {
                VerifyLandlordStep.Start -> {
                    _state.update { it.copy(currentStep = VerifyLandlordStep.Details) }
                }
                VerifyLandlordStep.Details -> viewModelScope.launch { submit() }
            }
        }

        override fun onSecondary() = Unit

        // MARK: - Field mutations

        fun setOwnerName(value: String) = updateForm { it.copy(ownerName = value) }

        fun setContactName(value: String) = updateForm { it.copy(contactName = value) }

        fun setEmail(value: String) = updateForm { it.copy(email = value) }

        fun setPhone(value: String) = updateForm(revalidate = false) { it.copy(phone = value) }

        fun setLease(lease: VerifyLandlordLeaseFile?) = updateForm { it.copy(lease = lease) }

        fun setPMEnabled(enabled: Boolean) =
            updateForm { current ->
                if (enabled) {
                    current.copy(pmEnabled = true)
                } else {
                    current.copy(pmEnabled = false, pmName = "", pmEmail = "", pmPhone = "")
                }
            }

        fun setPMName(value: String) = updateForm { it.copy(pmName = value) }

        fun setPMEmail(value: String) = updateForm { it.copy(pmEmail = value) }

        fun setPMPhone(value: String) = updateForm(revalidate = false) { it.copy(pmPhone = value) }

        /**
         * Used by previews / sample-data toggles + the dashboard
         * fast-track decision tree. Mirrors iOS' `setVariant(_:)`.
         */
        fun setVariant(variant: VerifyLandlordVariant) {
            val next =
                when (variant) {
                    VerifyLandlordVariant.Canonical -> VerifyLandlordSampleData.canonical
                    VerifyLandlordVariant.FastTrack -> VerifyLandlordSampleData.fastTrack
                }
            _state.update { it.copy(startContent = next) }
        }

        fun acknowledgeEvent() {
            pendingEvent.value = null
        }

        // MARK: - Submit

        @Suppress("ReturnCount")
        private suspend fun submit() {
            val snapshot = _state.value
            if (snapshot.isSubmitting) return
            val live = snapshot.form.validate()
            _state.update { it.copy(errors = live) }
            if (!live.isEmpty) {
                val noun = if (live.count == 1) "thing" else "things"
                _state.update {
                    it.copy(
                        submitState = VerifyLandlordSubmitState.Error("Fix ${live.count} $noun to submit"),
                    )
                }
                return
            }
            _state.update { it.copy(submitState = VerifyLandlordSubmitState.Submitting) }
            if (!networkMonitor.isOnline.value) {
                _state.update {
                    it.copy(
                        submitState =
                            VerifyLandlordSubmitState.Error(
                                "You're offline. Try again when you're back online.",
                            ),
                    )
                }
                return
            }
            // The landlord / property-manager details collected above have
            // no backend representation today — `request-postcard` takes no
            // body — so finishing the wizard simply mails the verification
            // postcard and routes to code entry. The details stay client-side.
            if (submitDelayMillis > 0) delay(submitDelayMillis)
            when (val result = verificationRepository.requestPostcard(homeId)) {
                is NetworkResult.Success -> {
                    _state.update { it.copy(submitState = VerifyLandlordSubmitState.Submitted) }
                    pendingEvent.value = VerifyLandlordOutboundEvent.OpenPostcardVerification(homeId)
                }
                is NetworkResult.Failure -> {
                    // A pending/duplicate code (400) or address cap (429)
                    // means a postcard is already on its way — proceed to
                    // enter it. Other failures surface inline so the user
                    // can retry.
                    if (result.error.code == 400 || result.error.code == 429) {
                        _state.update { it.copy(submitState = VerifyLandlordSubmitState.Submitted) }
                        pendingEvent.value = VerifyLandlordOutboundEvent.OpenPostcardVerification(homeId)
                    } else {
                        _state.update {
                            it.copy(
                                submitState =
                                    VerifyLandlordSubmitState.Error(
                                        result.error.message.ifEmpty {
                                            "Couldn't request the verification postcard. Try again."
                                        },
                                    ),
                            )
                        }
                    }
                }
            }
        }

        // MARK: - Chrome derivation

        private fun computeChrome(state: VerifyLandlordUiState): WizardChrome =
            when (state.currentStep) {
                VerifyLandlordStep.Start ->
                    WizardChrome(
                        title = "Verify landlord",
                        progressLabel = WizardProgressLabel.StepOf(1, TOTAL_STEPS),
                        progressFraction = 1f / TOTAL_STEPS,
                        leading = WizardLeadingControl.Close,
                        primaryCtaLabel = "Start verification",
                        primaryCtaEnabled = true,
                        secondaryCta = null,
                        isSubmitting = false,
                        dirty = state.isDirty,
                        showsProgressBar = true,
                    )
                VerifyLandlordStep.Details -> {
                    val live = state.form.validate()
                    val blocked = (state.errors != null && !live.isEmpty) || state.isSubmitting
                    WizardChrome(
                        title = "Verify landlord",
                        progressLabel = WizardProgressLabel.StepOf(2, TOTAL_STEPS),
                        progressFraction = 2f / TOTAL_STEPS,
                        leading = WizardLeadingControl.Back,
                        primaryCtaLabel = "Submit",
                        primaryCtaEnabled = !blocked,
                        secondaryCta = null,
                        isSubmitting = state.isSubmitting,
                        dirty = state.isDirty,
                        showsProgressBar = true,
                    )
                }
            }

        // MARK: - Helpers

        private inline fun updateForm(
            revalidate: Boolean = true,
            crossinline transform: (VerifyLandlordForm) -> VerifyLandlordForm,
        ) {
            _state.update { current ->
                val nextForm = transform(current.form)
                val nextErrors =
                    if (revalidate && current.errors != null) {
                        nextForm.validate()
                    } else {
                        current.errors
                    }
                current.copy(form = nextForm, errors = nextErrors)
            }
        }

        companion object {
            /** Total steps surfaced to the user — the third is A12.7
             *  (the sibling postcard verification screen). */
            const val TOTAL_STEPS: Int = 3
            const val SUBMIT_DELAY_DEFAULT_MILLIS: Long = 800L
        }
    }
